File: old/sorting/merge_sort/merge_sort_implementation/merge_sort_implementation.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> merge(vector<int>& sortedLeftArr, vector<int>& sortedRightArr) {
    vector<int> outputArr;
    int leftIndex = 0;
    int rightIndex = 0;

    while (leftIndex < sortedLeftArr.size() && rightIndex < sortedRightArr.size()) {
       if (sortedLeftArr[leftIndex] < sortedRightArr[rightIndex]) {
            outputArr.push_back(sortedLeftArr[leftIndex]);
            leftIndex += 1;
        } else {
            outputArr.push_back(sortedRightArr[rightIndex]);
            rightIndex += 1;
        } 
    }

    while (leftIndex < sortedLeftArr.size()) {
        outputArr.push_back(sortedLeftArr[leftIndex]);
        leftIndex += 1;    
    }


    while (rightIndex < sortedRightArr.size()) {
        outputArr.push_back(sortedRightArr[rightIndex]);
        rightIndex += 1;    
    }

    return outputArr;
}

vector<int> mergeSort(vector<int>& arr) {
    if (arr.size() <= 1) {
        return arr;
    }

    int middleIndex = (arr.size() / 2);

    vector<int> leftArr(arr.begin(), arr.begin() + middleIndex);
    vector<int> rightArr(arr.begin() + middleIndex, arr.end());

    vector<int> sortedLeftArr = mergeSort(leftArr);
    vector<int> sortedRightArr = mergeSort(rightArr);
    return merge(sortedLeftArr, sortedRightArr);
}

int main() {
    vector<int> numbers = {5, 3, 8, 4, 2, 7, 1, 6};
    vector<int> sortedNumbers = mergeSort(numbers);

    cout << "Sorted: ";
    for (int num : sortedNumbers) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: old/sorting/merge_sort/merge_sort_implementation/merge_sort_implementation.py
def merge(sorted_left_array: list[int], sorted_right_array) -> list[int]:
    sorted_list = []
    left_index = 0
    right_index = 0

    # merge into the sorted_list until we get out of the loop from the sorted left or sorted right or both
    while left_index < len(sorted_left_array) and right_index < len(sorted_right_array):
        if sorted_left_array[left_index] < sorted_right_array[right_index]:
            sorted_list.append(sorted_left_array[left_index])
            left_index += 1
        else:
            sorted_list.append(sorted_right_array[right_index])
            right_index += 1

    # if we still have things left in the sorted left array
    while left_index < len(sorted_left_array):
        sorted_list.append(sorted_left_array[left_index])
        left_index += 1
    while right_index < len(sorted_right_array):
        sorted_list.append(sorted_right_array[right_index])
        right_index += 1

    return sorted_list


def merge_sort(numbers: list[int]) -> list[int]:
    # if the list contains only one number return the list it is already sorted
    if len(numbers) <= 1:
        return numbers

    # else find the middle index
    middle_index = len(numbers) // 2

    # sort the left array upto middle index and right array from middle index to end and merge the sorted left and right arrays
    sorted_left_array = merge_sort(numbers[:middle_index])
    sorted_right_array = merge_sort(numbers[middle_index:])
    return merge(sorted_left_array, sorted_right_array)


def main():
    sorted_list = merge_sort([1, 3, 5, 2, 4, 6])
    print(sorted_list)


main()
